src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs

```diff
@@ -176,8 +176,15 @@ private void EnsureApplicationServices()
         {
             if (_applicationServices == null)
             {
+                _applicationServices = (IServiceProvider)_applicationServiceCollection
+                    .LastOrDefault(d => d.ServiceType == typeof(IServiceProvider) &&
+                                        d.ImplementationInstance != null)?
+                    .ImplementationInstance;
```

Comment:
We can not do this: ``_hostingServiceProvider.GetService<IServiceProvider>();`` because it will return _hostingServiceProvider. We need to look it up in the list.

Reply:
I think the entire feature is undesirable really and this is a hack.

Reply:
The problem is that there are 3 service providers and the service provider you get when you build is different to the one added to the IServiceCollection. It's using the IServiceCollection like a property bag to shuttle the information to the WebHost